chain/consensus/mir/mine.go

```diff
@@ -0,0 +1,185 @@
+	if err != nil {
+		return err
+	}
+	mirErrors := mirAgent.Start(ctx)
+	mirHead := mirAgent.App.ChainNotify
+
+	netName, nerr := api.StateNetworkName(ctx)
+	if nerr != nil {
+		return nerr
+	}
+	subnetID := address.SubnetID(netName)
+
+	log.Infof("Mir miner params: network name - %s, subnet ID - %s", netName, subnetID)
+
+	submit := time.NewTicker(SubmitInterval)
+	defer submit.Stop()
+
+	for {
+		base, err := api.ChainHead(ctx)
+		if err != nil {
+			log.Errorw("failed to get the head of chain", "error", err)
+			continue
+		}
+
+		select {
+		case <-ctx.Done():
+			log.Debug("Mir miner: context closed")
+			return nil
+		case merr := <-mirErrors:
+			return merr
+		case mb := <-mirHead:
+			log.Debugf(">>>>> received %d messages in Mir block", len(mb))
+			msgs, crossMsgs := getMessagesFromMirBlock(mb)
+
+			log.Infof("[subnet: %s, epoch: %d] try to create a block", subnetID, base.Height()+1)
+			bh, err := api.MinerCreateBlock(ctx, &lapi.BlockTemplate{
+				// TODO: if there are several nodes then the miner is the Mir node proposed the block within Mir.
+				Miner:            miner,
+				Parents:          base.Key(),
+				BeaconValues:     nil,
+				Ticket:           &ltypes.Ticket{VRFProof: nil},
+				Epoch:            base.Height() + 1,
+				Timestamp:        uint64(time.Now().Unix()),
+				WinningPoStProof: nil,
+				Messages:         msgs,
+				CrossMessages:    crossMsgs,
+			})
+			if err != nil {
+				log.Errorw("creating block failed", "error", err)
+				continue
+			}
+			if bh == nil {
+				log.Debug("created a nil block")
+				continue
+			}
+
+			log.Infof("[subnet: %s, epoch: %d] try to sync a block", subnetID, base.Height()+1)
+
+			err = api.SyncSubmitBlock(ctx, &types.BlockMsg{
+				Header:        bh.Header,
+				BlsMessages:   bh.BlsMessages,
+				SecpkMessages: bh.SecpkMessages,
+				CrossMessages: bh.CrossMessages,
+			})
+			if err != nil {
+				log.Errorw("unable to sync a block", "error", err)
+				continue
+			}
+
+			log.Infof("[subnet: %s, epoch: %d] mined a block %v", subnetID, bh.Header.Height, bh.Cid())
+		case <-submit.C:
+			msgs, err := api.MpoolSelect(ctx, base.Key(), 1)
+			if err != nil {
+				log.Errorw("unable to select messages from mempool", "error", err)
+			}
+			log.Debugf("[subnet: %s, epoch: %d] retrieved %d msgs", subnetID, base.Height()+1, len(msgs))
+
+			crossMsgs, err := api.GetUnverifiedCrossMsgsPool(ctx, subnetID, base.Height()+1)
+			if err != nil {
+				log.Errorw("unable to select cross-messages from mempool", "error", err)
+			}
+			log.Debugf("[subnet: %s, epoch: %d] retrieved %d crossmsgs", subnetID, base.Height()+1, len(crossMsgs))
+
+			for _, msg := range msgs {
+				msgBytes, err := msg.Serialize()
+				if err != nil {
+					log.Error("unable to serialize message:", err)
+					continue
+				}
+
+				reqNo := mirTypes.ReqNo(msg.Message.Nonce)
+				tx := common.NewSignedMessageBytes(msgBytes, nil)
+
+				// TODO: define what client ID is in Eudico case.
```

Comment:
Yes. I'll shy away of requiring different IDs. The right interface in MIR may do the trick.

Reply:
If I understood correctly from previous discussions, each message has a "nonce", which is a monotonically increasing number (I'd prefer calling it a "message number", "request number", or "sequence number", but as long as we all know what we're talking about, I don't care that much). This nonce uniquely identifies a message among all the messages coming, in a way, from the same "source". This source is something that produces a sequence of messages and assigns them the nones. In other words, the source defines a set of messages among which nonces are unique. The identifier of this source is what we need to use as "client ID". What is this identifier in Eudico terms? Actor address? Wallet address?

Reply:
We discussed with @adlrocha : subnetName + wallet address can be used as IDs.